ocaml/lib/pickles_bindings.ml

```diff
@@ -368,6 +369,134 @@ module Choices = struct
     get_rules (Choices (fun ~self:_ -> [])) (Array.length js_rules - 1)
 end
 
+module Cache = struct
+  module Sync : Key_cache.Sync = struct
+    open Key_cache
+    include T (Or_error)
+
+    module Disk_storable = struct
+      include Disk_storable (Or_error)
+
+      let of_binable = Trivial.Disk_storable.of_binable
+
+      let simple to_string read write = { to_string; read; write }
+    end
+
+    let read spec { Disk_storable.to_string; read; write = _ } key =
+      Or_error.find_map_ok spec ~f:(fun s ->
+          let res, cache_hit =
+            match s with
+            | Spec.On_disk { should_write; _ } ->
+                let path = to_string key in
+                ( read ~path key
+                , if should_write then `Locally_generated else `Cache_hit )
+            | S3 _ ->
+                (Or_error.errorf "Downloading from S3 is disabled", `Cache_hit)
+          in
+          Or_error.map res ~f:(fun res -> (res, cache_hit)) )
+
+    let write spec { Disk_storable.to_string; read = _; write } key value =
+      let errs =
+        List.filter_map spec ~f:(fun s ->
+            let res =
+              match s with
+              | Spec.On_disk { should_write; _ } ->
+                  if should_write then write key value (to_string key)
+                  else Or_error.return ()
+              | S3 _ ->
+                  Or_error.return ()
+            in
+            match res with Error e -> Some e | Ok () -> None )
+      in
+      match errs with [] -> Ok () | errs -> Error (Error.of_list errs)
+  end
+
+  let () = Key_cache.set_sync_implementation (module Sync)
+
+  open Pickles.Cache
+
+  type any_key =
```

Comment:
That seems to be code that can be refactorised, but I don't have any idea off the top of my head.

Reply:
we could pass a separate storage implementation from JS for every type of prover/verifier key -- then we wouldn't need those enums. but it would make the OCaml-TS interface much bigger. I prefer keeping the interface as small as possible because it is the most brittle part of the system (no type guarantees etc), that's why I went with a single `storable` that is passed from JS, which handles all key types at once